### DISCUSSION TITLE
I can't open a file in /data

### DISCUSSION THREAD

**Discussion by mimimix** [zepp-health/zeppos-forum#240] [Q&A] [answered]
I just need to read the file test.txt , but when I try to get a file or the contents of a folder, I always get undefined. What I'm doing wrong :(

![folders](https://github.com/zepp-health/zeppos-forum/assets/39842497/cf94d289-64ce-4118-9425-af52966bbc81)
![code](https://github.com/zepp-health/zeppos-forum/assets/39842497/e1daaa5a-436d-48fc-a77c-120fb3e10659)
![output](https://github.com/zepp-health/zeppos-forum/assets/39842497/20b1d523-69e7-4aab-914b-821fe27bfd6c)

![simulator version (GTS 4 v1.0.5)](https://github.com/zepp-health/zeppos-forum/assets/39842497/5e398936-379c-4145-a996-adc658789483)

**Comment by silver-zepp** [comment]
Refer to the hierarchy on this doc: https://docs-testing.zepp.com/docs/guides/framework/device/fs/

It looks to me that your **data** folder is stored outside the **assets** folder. In the future please make sure you attach a reproducible project. It can get very tricky to debug a screenshot. @Mimimix

**Reply by mimimix** [reply]
[fileTest.zip](https://github.com/zepp-health/zeppos-forum/files/12516530/fileTest.zip)

**Comment by silver-zepp** [comment, marked as answer]
I see now. Here's a working example for you to play with: 
```js
import { openAssetsSync, readSync, } from "@zos/fs";

const FILE_PATH = "data/test.txt";

function bufferToString(buffer){
  return String.fromCharCode.apply(null, new Uint8Array(buffer));
}

Page({
  build() {
    const fd = openAssetsSync({
      path: FILE_PATH,
      flag: O_RDONLY,
    });

    const buffer = new ArrayBuffer(1024);
    const result = readSync({
      fd,
      buffer,
    });

    if (result > 0) {
      console.log("readSync succeed");
      console.log(bufferToString(buffer));
    } else {
      console.log("readSync failed", result);
    }
  },
});
```

Your hierarchy for this project should be: `assets/390x450-gts-4/data/test.txt`
[fileTest-fixed.zip](https://github.com/zepp-health/zeppos-forum/files/12516770/fileTest-fixed.zip)

**Reply by silver-zepp** [reply]
To elaborate on the issue that you faced - reading an asset is a bit more challenging. The methods you've used are designed to work with the files that you create and then read on the device. 

But in this case, you are providing your own asset that is contained in the same context as your app. The `/data` folder is not located inside the app folder. It has its own location on the device. Think of it as an `%appdata%` folder in `Windows`.

**Reply by mimimix** [reply]
I downloaded the folder with your code and tried to run it, but I get the error "readSync failed -2"
Node.js v20.4.0
Zeus version: 1.2.3

**Reply by silver-zepp** [reply]
Try implementing the code I've provided into your original folder. Also, it looks like I've removed the read-only flag when cleaning up the code. Try adding it back:

```js
import { openAssetsSync, readSync, O_RDONLY } from "@zos/fs";
// as well as the result can be >=
if (result >= 0) {
// ...
}
```

For me, it compiles, either way, might be using cache or pulling it from the interface. Also, you can try updating zeus with:
```
npm i -g @zeppos/zeus-cli@1.3.0-beta.12
```
If that doesn't work, you can roll back to your current. Let me know if you have any progress.

**Reply by silver-zepp** [reply]
![image](https://github.com/zepp-health/zeppos-forum/assets/138894963/8ec064e4-0b83-4ed5-ac68-ce2a83620064)

**Reply by mimimix** [reply]
After updating zeus-cli to 1.3.0-beta.12 everything worked.
But sometimes an error occurs at startup (also relevant for the previous version of zeus):

```
info  connecting to simulator
success  simulator connected
rebuilding...
info  Start building package, targets: 390x450-gts-4.
node:internal/assert:14
    throw new ERR_INTERNAL_ASSERTION(message);
    ^

Error [ERR_INTERNAL_ASSERTION]: This is caused by either a bug in Node.js or incorrect usage of Node.js internals.
Please open an issue with this stack trace at https://github.com/nodejs/node/issues

    at new NodeError (node:internal/errors:405:5)
    at assert (node:internal/assert:14:11)
    at internalConnectMultiple (node:net:1118:3)
    at Timeout.internalConnectMultipleTimeout (node:net:1687:3)
    at listOnTimeout (node:internal/timers:575:11)
    at process.processTimers (node:internal/timers:514:7) {
  code: 'ERR_INTERNAL_ASSERTION'
}

Node.js v20.6.0
```
But it starts 2-3 times

**Reply by silver-zepp** [reply]
I'd say it has something to do with node rather than zeus, never seen this issue, and considering you've met it with both versions it's most likely a node.js issue. Maybe a corruption. But it also might be that it's not supported. I haven't heard anyone having this before and people tend to use the most recent versions. Weird. 

I personally still use `v16.14.0`. Try installing the LTS version?

**Reply by mimimix** [reply]
When I installed node 18.17.1, the error disappeared :)
Thanks for the help

**Reply by silver-zepp** [reply]
Awesome! Glad we got it all sorted. Happy hacking @Mimimix 👋

**Comment by SashaCX75** [comment]
I also had problems with reading files. 
I saved the file using hmFS.write and then read it again using hmFS.read. Everything works until I restart the programme in the simulator. That is, when I make changes to the programme and run it again in the simulator, the previously saved file disappears. On the watch everything works correctly and the file does not disappear.

**Reply by silver-zepp** [reply]
That's a flaw in VM's design. When you work with the filesystem on the Simulator and you want to interact with files, you need to make sure that you don't push any updates on the VM. In other words, do not restart the VM by re-saving the project in the `VSCode`. This way destroys the `/data` folder and all its contents.

You will have to exit the app with a `Left -> Right swipe` or press the `Home` button, and then enter into it again to have your `FS` changes persistent @SashaCX75

**Comment by SashaCX75** [comment]
@silver-zepp This only partially solves the problem. While I'm working with one version of the program, this method works. But as soon as I update the program, the saved data will still disappear.

It would not be bad to add such a feature to the description of the operation of these functions when working on a simulator. I spent a lot of time thinking that I was not storing data correctly and trying to solve this problem.

**Reply by silver-zepp** [reply]
I hear you @SashaCX75. I reported this 4 months ago as well, but it never got into the right hands or was lost in translation. My suggestion was to add an option in settings to clear the storage only when required. I'll make sure that this time it will be heard.

**Reply by silver-zepp** [reply]
This was fixed for multiple devices. Look for the VM update with the remark "Improve Mini Program installation speed". If it has it then the storage should be properly working @SashaCX75 
![image](https://github.com/zepp-health/zeppos-forum/assets/138894963/e1f82dfa-0280-46eb-89c6-d742028db2f3)

**Reply by SashaCX75** [reply]
@silver-zepp Thank you. In versions of the simulator where there is such a note it is corrected. But it seems to be only for models on ZeppOS 2 and newer. And I was just working with GTS 3 and it's not fixed there yet. I hope it is temporary.

I would also like the screen outline to be greyed out like in Balance simulator. It is very convenient, you can see the screen borders at once.

**Comment by oakeley** [comment]
@silver-zepp I am trying to communicate with the phone via an http connection to 127.0.0.1. The watchface uses a miniapp only as a bridge. The miniapp writes the http results to a file that the watchface then loads. Can you recommend an alternative method for getting data from the phone to the watchface?
In Zepp v1 and Zepp v2 the following command in the watchface works:
**const file = hmFS.open(filename, hmFS.O_RDONLY);**
I have tried this but it still doesn't work:
const file = hmFS.open(filename, hmFS.O_RDONLY, {appid: 28962});

Many thanks for your help

**Reply by oakeley** [reply]
Is there a page somewhere that lists all of the @zos/fs features? Specifically it would be super helpful for developers if you could expose an "autocomplete" for vscode that includes the @zos capabilities... Thanks
